src/monitor/health_check_worker.c

```diff
@@ -262,21 +264,59 @@ HealthCheckWorkerLauncherMain(Datum arg)
 															HASH_ENTER, &isFound);
 			if (isFound && dbData->isActive)
 			{
-				/* This database has already been processed */
+				handle = dbData->handle;
+
 				LWLockRelease(&HealthCheckHelperControl->lock);
+
+				/*
+				 * This database has already been processed.
+				 *
+				 * Perform a quick and inexpensive check to verify that it is
+				 * actually running. Note that it is not possible to get
+				 * BGWH_NOT_YET_STARTED at this point, because this is not first
+				 * time we try to register the worker due to the isFound value
+				 * above, and we will WaitForBackgroundWorkerStartup after
+				 * registering before setting isActive to true. Thus we can only
+				 * get BGWH_STARTED or BGWH_STOPPED.
+				 *
+				 * NOTE: For PostgreSQL versions prior to 12, it is possible to
+				 * get erroneous results in GetBackgroundWorkerPid due to the
+				 * slot generation number being out of date. This is especially
+				 * true after signalling the Postmaster to reload the
+				 * configuration.
+				 */
+				if (GetBackgroundWorkerPid(handle, &pid) != BGWH_STARTED)
+				{
+					dbData->isActive = false;
+					#if (PG_VERSION_NUM >= 120000)
+					ereport(WARNING,
+							(errmsg(
+								 "found stopped worker for pg_auto_failover "
+								 "health checks in \"%s\"",
+								 entry->dbname)));
+					#else
+					ereport(WARNING,
+							(errmsg(
+								 "stopping worker for pg_auto_failover "
+								 "health checks in \"%s\"",
+								 entry->dbname)));
+					#endif
+
+					StopHealthCheckWorker(entry->dboid);
```

Comment:
Can you add a comment about we stop the worker here? I can guess why, but I'm not sure, and would appreciate being able to read the reasoning here.